### ISSUE TITLE
analyze code: convert_checkpoint_gpu_to_cpu.lua

### ISSUE THREAD

**Issue by blueberrynfig** [blueberrynfig/illegal-dumping-object-detection-engine#9]

**Comment by MalvinMiao** [comment]
This is a script for converting GPU checkpoint to CPU checkpoint.

CPU checkpoints are not saved by the training script automatically because of Torch cloning limitations.

command line:

- model
- option: -gpuid (0 means GPU, -1 means CPU)

torch.manualSeed( ) :The random number generator is provided with a random seed via seed() when torch is being initialized. It can be reinitialized using seed() or manualSeed().

A random seed (or seed state, or just seed) is a number (or vector) used to initialize a pseudorandom number generator.

```
function torch.setdefaulttensortype(typename)

assert(type(typename) == 'string', 'string expected')

if torch.getconstructortable(typename) then

torch.Tensor = torch.getconstructortable(typename)

torch.Storage = torch.getconstructortable(torch.typename(torch.Tensor(1):storage()))

else

error(string.format("<%s> is not a string describing a torch object", typename))

end

end
```

The Tensor class is probably the most important class in Torch. Almost every package depends on this class. It is the class for handling numeric data. As with pretty much anything in Torch7, tensors are serializable.

A Tensor is a potentially multi-dimensional matrix. The number of dimensions is unlimited that can be created using LongStorage with more dimensions.

checkpoint = model
https://karpathy.github.io/2015/05/21/rnn-effectiveness/
checkpoint.protos sotres all the networks

cudnnNetToCpu(net) calls replaceModules(net, orig_class_name, replacer) to replace original model with **nn.SpatialConvolutional** and **nn.SpatialMaxPooling**.

A convolution is an integral that expresses the amount of overlap of one function g as it is shifted over another function f.

**SpatialConvolutional**:
Applies a 2D convolution over an input image composed of several input planes. The input tensor in forward(input)is expected to be a 3D tensor (nInputPlane x height x width).

**SpatialMaxPooling**:
Pooling is a way of sub-sampling, i.e. reducing the dimension of the input (or at some hidden layer)

assume a group size of 2X2 on the following 4X4 matrix

4 5 | 4 5
6 7 | 8 7
_____
2 3 | 6 1
9 6 | 5 4

then,
1) Max pooling: takes the max out of each group:
so the matrix after max pooling will be:
7 | 8
------
9 | 6

2) Avg pooling: takes the average of each group
so the matrix after avg pooling will be:
5.50 | 6
--------
5.25 | 4

Applies 2D max-pooling operation in kWxkH regions by step size dWxdH steps. The number of output features is equal to the number of input planes.
If the input image is a 3D tensor nInputPlane x height x width, the output image size will be nOutputPlane x oheight x owidth

nn is the Neural Network Package of Torch.
